File: Data Structures/Queue/Java/Queue.java
public class Queue <T>{
    private T[] queueArray;
    private int rear, front;
    private int count;
    private final int DEFAULT_SIZE = 50;

    /*
    @SuppressWarnings("unchecked") annotations are used to prevent the "Unchecked cast" error
    */

    // Default Constructor
    @SuppressWarnings("unchecked")
    public Queue(){
        queueArray = (T[]) new Object[DEFAULT_SIZE];

        // rear and front variables represent their index
        this.rear = -1;
        this.front = 0;
        // count variable represents the element count in the array
        this.count = 0;
    }

    // Parameterized Constructor with Specific Size
    @SuppressWarnings("unchecked")
    public Queue(int initialSize){
        queueArray = (T[]) new Object[initialSize];
        this.rear = 0;
        this.front = 0;
        this.count = 0;
    }

    // Enqueue method for adding an element to the rear of a queue (FIFO)
    public void enqueue(T data){
        if (isFull()){
            extend();
        }

        rear++;
        queueArray[rear] = data;
        count++;
    }

    // Dequeue method for removing an element from the front of a queue (FIFO)
    public T dequeue(){
        if (isEmpty()){
            throw new IllegalStateException("Queue is empty. Dequeue cannot performed.");
        }

        T item = queueArray[front];
        queueArray[front] = null;
        front++;
        count--;

        // Checking if the queue has become significantly smaller (25% of its capacity) compared to its length
        if (count == queueArray.length / 4){
            shrink();
        }

        return item;
    }

    // Peek method for getting value of the front element
    public T peek(){
        return queueArray[front];
    }

    private boolean isFull(){
        return count == queueArray.length;
    }

    private boolean isEmpty(){
        return count == 0;
    }

    public void printQueue(){
        if (isEmpty()){
            System.out.println("Queue is empty.");
            return;
        }

        for (int i = front; i <= rear; i++)
        {
            System.out.println(queueArray[i]);
        }
    }

    @SuppressWarnings("unchecked")
    public void clear() {
        front = 0;
        rear = -1;
        count = 0;

        T[] newArray = (T[]) new Object[queueArray.length];
        System.arraycopy(queueArray, 0, newArray, 0, count);
        queueArray = newArray;
    }

    // Since count is declared as private the method provides access to the private 'count' variable
    public int count(){
        return count;
    }

    // Extend method for increasing the size of the underlying array, it prevents exceeding the range
    @SuppressWarnings("unchecked")
    private void extend(){
        T[] newArray = (T[]) new Object[queueArray.length * 2];
        System.arraycopy(queueArray, 0, newArray, 0, queueArray.length);
        queueArray = newArray;
    }

    // Shrink method for reducing the size of the underlying array
    @SuppressWarnings("unchecked")
    private void shrink(){
        T[] newArray = (T[]) new Object[queueArray.length / 2];
        System.arraycopy(queueArray, 0, newArray, 0, count);
        queueArray = newArray;
    }
}

File: Data Structures/Queue/Python/queue.py
class Queue:
    def __init__(self, initial_size=50):
        self.queue_arr = [None] * initial_size
        self.rear = -1
        self.front = 0
        self.count = 0

    # Enqueue function for adding an element to the rear of a queue (FIFO)
    def enqueue(self, data):
        if self.is_full():
            self.extend()

        self.rear += 1
        self.queue_arr[self.rear] = data
        self.count += 1

    # Dequeue function for removing an element from the front of a queue (FIFO)
    def dequeue(self):
        if self.is_empty():
            raise ValueError("Queue is empty. Dequeue cannot performed.")

        item = self.queue_arr[self.front]
        self.queue_arr[self.front] = None
        self.front += 1
        self.count -= 1

        # Checking if the queue has become significantly smaller (25% of its capacity) compared to its length
        if self.count == len(self.queue_arr) // 4:
            self.shrink()

        return item

    # Peek function for getting value of the front element
    def peek(self):
        return self.queue_arr[self.front]

    # Count function for getting element count in the stack
    def get_count(self):
        return self.count

    def is_full(self):
        return self.count == len(self.queue_arr)

    def is_empty(self):
        return self.count == 0

    def print_queue(self):
        if self.is_empty():
            print("Queue is empty.")
            return

        for i in range(self.front, self.rear + 1):
            print(self.queue_arr[i])

    # Extend function for increasing the size of the underlying array, it prevents exceeding the range
    def extend(self):
        new_array = [None] * (len(self.queue_arr) * 2)
        new_array[:len(self.queue_arr)] = self.queue_arr
        self.queue_arr = new_array

    # Shrink function for reducing the size of the underlying array
    def shrink(self):
        new_array = [None] * (len(self.queue_arr) // 2)
        new_array[:self.front + 1] = self.queue_arr[:self.front + 1]
        self.queue_arr = new_array
